src/FontManager.cc

```diff
@@ -1,251 +1,265 @@
-long findFonts(ResultSet **, FontDescriptor *);
-long findFont(FontDescriptor **, FontDescriptor *);
-long substituteFont(FontDescriptor **, char *, char *);
+using namespace Napi;
 
 // converts a ResultSet to a JavaScript array
-Local<Array> collectResults(ResultSet *results) {
-  Nan::EscapableHandleScope scope;
-  Local<Array> res = Nan::New<Array>(results->size());
+Napi::Array collectResults(Napi::Env env, ResultSet *results) {
+  Napi::EscapableHandleScope scope(env);
+  Napi::Array res = Napi::Array::New(env, results->size());
 
   int i = 0;
   for (ResultSet::iterator it = results->begin(); it != results->end(); it++) {
-    Nan::Set(res, i++, (*it)->toJSObject());
+    res.Set(i++, (*it)->toJSObject(env));
   }
 
   delete results;
-  return scope.Escape(res);
+  return scope.Escape(res).As<Napi::Array>();
 }
 
 // converts a FontDescriptor to a JavaScript object
-Local<Value> wrapResult(FontDescriptor *result) {
-  Nan::EscapableHandleScope scope;
+Napi::Value wrapResult(Napi::Env env, FontDescriptor *result) {
+  Napi::EscapableHandleScope scope(env);
   if (result == NULL)
-    return scope.Escape(Nan::Null());
+    return scope.Escape(env.Null());
 
-  Local<Object> res = result->toJSObject();
+  Napi::Object res = result->toJSObject(env);
   delete result;
   return scope.Escape(res);
 }
 
-// holds data about an operation that will be
-// performed on a background thread
-struct AsyncRequest {
-  uv_work_t work;
-  FontDescriptor *desc;     // used by findFont and findFonts
-  char *postscriptName;     // used by substituteFont
-  char *substitutionString; // ditto
-  FontDescriptor *result;   // for functions with a single result
-  ResultSet *results;       // for functions with multiple results
-  long error;               // result/results is defined if error == 0
-  Nan::Callback *callback;  // the actual JS callback to call when we are done
-
-  AsyncRequest(Local<Value> v) {
-    work.data = (void *)this;
-    callback = new Nan::Callback(v.As<Function>());
-    desc = NULL;
-    postscriptName = NULL;
-    substitutionString = NULL;
-    result = NULL;
-    results = NULL;
-  }
-
-  ~AsyncRequest() {
-    delete callback;
-
-    if (desc)
-      delete desc;
+// handles asynchrnous execution for interface functions
+// this destroys itself unless we call AsyncWorker::SuppressDestruct()
+template<typename _ReturnType>
+struct PromiseAsyncWorker : public Napi::AsyncWorker {
+  static_assert(std::is_same<_ReturnType, ResultSet>::value || std::is_same<_ReturnType, FontDescriptor>::value, "");
+
+  PromiseAsyncWorker(Napi::Env env, std::function<long(_ReturnType **)> fn) :
+    Napi::AsyncWorker(env),
+    deferred(Napi::Promise::Deferred::New(env)), fn(fn) {}
+
+  virtual void Execute() override {
+    this->result = nullptr;
+    long error = this->fn(&this->result);
+    if(error) {
+      PromiseAsyncWorker::SetError(std::string("Function failed with error code ") + std::to_string(error));
```

Comment:
I wouldn't use scope resolution here, it might look like that `PromiseAsyncWorker` has a new overriding `SetError` method. I'd use directly `SetError`, since `::` is normally used to call static methods.

Reply:
You're right, I don't remember why I used this syntax :thinking:. Can't think of a good reason now.